akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala

```diff
@@ -36,240 +35,135 @@ private object PoolSlot {
+    val eventsOut: Outlet[RawSlotEvent] = Outlet("SlotProcessor.eventsOut")
 
-    override def requestStrategy = ZeroRequestStrategy
+    override def shape: FanOutShape2[SlotCommand, ResponseContext, RawSlotEvent] = new FanOutShape2(in, responsesOut, eventsOut)
 
-    /**
-     * How PoolProcessor changes its `receive`:
-     * waitingExposedPublisher -> waitingForSubscribePending -> unconnected ->
-     * waitingForDemandFromConnection OR waitingEagerlyConnected -> running
-     * Given slot can become get to 'running' state via 'waitingForDemandFromConnection' or 'waitingEagerlyConnected'.
-     * The difference between those two paths is that the first one is lazy - reacts to DispatchCommand and then uses
-     * inport and outport actors to obtain more items.
-     * Where the second one is eager - reacts to SlotShouldConnectCommand from PoolConductor, sends SlotEvent.ConnectedEagerly
-     * back to conductor and then waits for the first DispatchCommand
-     */
-    override def receive = waitingExposedPublisher
+    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with InHandler {
+      private var firstRequest: RequestContext = _
+      private val inflightRequests = new java.util.ArrayDeque[RequestContext]()
 
-    def waitingExposedPublisher: Receive = {
-      case ExposedPublisher(publisher) ⇒
-        exposedPublisher = publisher
-        context.become(waitingForSubscribePending)
-      case other ⇒ throw new IllegalStateException(s"The first message must be `ExposedPublisher` but was [$other]")
-    }
+      private var connectionFlowSource: SubSourceOutlet[HttpRequest] = _
+      private var connectionFlowSink: SubSinkInlet[HttpResponse] = _
 
-    def waitingForSubscribePending: Receive = {
-      case SubscribePending ⇒
-        exposedPublisher.takePendingSubscribers() foreach (s ⇒ self ! ActorPublisher.Internal.Subscribe(s))
-        log.debug("become unconnected, from subscriber pending")
-        context.become(unconnected)
-    }
+      private var isConnected = false
 
-    val unconnected: Receive = {
-      case OnNext(DispatchCommand(rc: RequestContext)) ⇒
-        val (connInport, connOutport) = runnableGraph.run()
-        connOutport ! Request(totalDemand)
-        context.become(waitingForDemandFromConnection(connInport = connInport, connOutport = connOutport, rc))
+      def disconnect(ex: Option[Throwable] = None) = {
+        connectionFlowSource.complete()
+        if (isConnected) {
+          isConnected = false
 
-      case OnNext(ConnectEagerlyCommand) ⇒
-        val (in, out) = runnableGraph.run()
-        onNext(SlotEvent.ConnectedEagerly(slotIx) :: Nil)
-        out ! Request(totalDemand)
-        context.become(waitingEagerlyConnected(connInport = in, connOutport = out))
+          val (inflightRetry, inflightFail) = if (ex.isEmpty) (inflightRequests.asScala, Nil)
```

Comment:
That's exactly the pipelining support of the lower layers in `OutgoingConnectionBlueprint` which will pull new requests even while another one is still in flight and send these out as well. So, in fact, a new request may or may not been have sent out. As soon as you push it to the connection it will try to send it out. This may happen concurrently with the server sending back the previous response and closing the connection.